### DISCUSSION TITLE
framework error when using Frame_trigger

### DISCUSSION THREAD

**Discussion by bdudok** [pyControl/code#96] [Q&A] [answered]
Hi,
 I have searched for this but haven't found anything. I added one line to my task file to start a Frame_trigger (for camera), and added the event name to the list of events. Right at start, the framework throws and error on task start, and then it's outputting the pulse train (and driving the camera correctly) until a new connection is made to the board. What may be the issue? The task works fine if I just comment out the line that adds the trigger. 

This error persists regardless of which output I use for this (BNC or DAC).
The error persists even if all other outputs are commented out.

This is the line: (see full file attached)
frame_trigger = Frame_trigger(pin=board.DAC_1, pulse_rate=30, name='frame_trigger')
[RandomForaging.py.txt](https://github.com/pyControl/code/files/12410876/RandomForaging.py.txt)

And this is the log:
Micropython version: 1.191
Framework version: 1.8.1

Transferring state machine RandomForaging to pyboard. OK

Run started at: 2023/08/21 18:10:59

P 0 frame_trigger outputting pulses at 30Hz
D 0 trial_start

Error: uncaught exception of type: <class 'KeyError'>

**Comment by alustig3** [comment]
Hi @bdudok 

Is there additional information in your `Errorlog.txt` file that you can provide?

**Reply by bdudok** [reply]
Yes, see below. Thanks for looking into this!
2023-08-21 18:11:00,058 Traceback (most recent call last):
File "C:\Users\User\PycharmProjects\PyControl\gui\run_task_tab.py", line 453, in update
self.board.process_data()
File "C:\Users\User\PycharmProjects\PyControl\com\pycboard.py", line 484, in process_data
self.data_logger.process_data(new_data)
File "C:\Users\User\PycharmProjects\PyControl\com\data_logger.py", line 75, in process_data
data_consumer.process_data(new_data)
File "C:\Users\User\PycharmProjects\PyControl\gui\plotting.py", line 77, in process_data
self.analog_plot.process_data(new_data)
File "C:\Users\User\PycharmProjects\PyControl\gui\plotting.py", line 237, in process_data
self.data[ID] = np.roll(self.data[ID], -new_len, axis=0)
KeyError: 9

**Comment by ThomasAkam** [comment, marked as answer]
Hi @bdudok ,

I looked into this and there was a bug in the analog plotting function which meant that a crash occured if a Frametrigger was used with annother type of analog input (in your case the rotary encoder).  I fixed it by changing the process_data method of the analog plot ([here](https://github.com/pyControl/code/blob/master/gui/plotting.py#L229-239)) and pushed the updated code to the master branch on github.  The bug is fixed in the development branch so will be fixed in the next release.

best,

Thomas

**Reply by bdudok** [reply]
Thanks so much, it works now!
Best,
Barna
